numpyro/contrib/nn/auto_reg_nn.py

```diff
@@ -57,3 +60,117 @@ def create_mask(input_dim, hidden_dims, permutation, output_dim_multiplier):
+
+    MADE: Masked Autoencoder for Distribution Estimation [arXiv:1502.03509]
+    Mathieu Germain, Karol Gregor, Iain Murray, Hugo Larochelle
+    """
+    def __init__(self, input_dim, hidden_dims, param_dims=[1, 1], nonlinearity=relu):
+        self.input_dim = input_dim
+        self.hidden_dims = hidden_dims
+        self.masked_layers = None
+        self.param_dims = param_dims
+        self.nonlinearity = nonlinearity
+        self.all_ones = (np.array(param_dims) == 1).all()
+
+        # Calculate the indices on the output corresponding to each parameter
+        ends = onp.cumsum(self.param_dims, axis=0)
+        starts = onp.concatenate((np.zeros(1), ends[:-1]))
+        self.param_slices = [slice(int(s), int(e)) for s, e in zip(starts, ends)]
+
+        self.count_params = len(self.param_dims)
+        self.output_multiplier = sum(self.param_dims)
+
+        # Hidden dimension must be not less than the input otherwise it isn't
+        # possible to connect to the outputs correctly
+        for h in self.hidden_dims:
+            if h < self.input_dim:
+                raise ValueError('Hidden dimension must not be less than input dimension.')
+
+    def init_fun(self, rng, input_shape, permutation=None):
+        """
+        :param rng: rng used to initialize parameters
+        :param input_shape: input shape
+        :param permutation: an optional permutation that is applied to the inputs and controls the order of the
+            autoregressive factorization. in particular for the identity permutation the autoregressive structure
+            is such that the Jacobian is upper triangular. By default this is chosen at random.
+        :type permutation: array of ints
+        """
+        if permutation is None:
+            # By default set a random permutation of variables, which is important for performance with multiple steps
+            rng, rng_perm = random.split(rng)
+            self.permutation = onp.array(random.shuffle(rng_perm, onp.arange(self.input_dim)))
+        else:
+            self.permutation = permutation
+
+        # Create masks (no skip connections allowed; TODO add support)
+        masks, _ = create_mask(input_dim=self.input_dim, hidden_dims=self.hidden_dims,
+                               permutation=self.permutation,
+                               output_dim_multiplier=self.output_multiplier)
+
+        # Create masked layers
+        self.masked_layers = [MaskedDense(mask) for mask in masks]
+        init_params = []
+        for mask in self.masked_layers:
+            mask_init = mask[0]
+            input_shape, param = mask_init(rng, input_shape)
+            init_params.append(param)
+
+        return input_shape, init_params
+
+    def apply_fun(self, params, inputs, **kwargs):
+        """
+        :param params: layer parameters
+        :param inputs: layer inputs
+        """
+        out = inputs
+
+        for k, (mask, param) in enumerate(zip(self.masked_layers, params)):
```

Comment:
I think this might need a few modifications to make it jittable, like using jax control flow primitives - using `fori_loop` instead of native for loop which will be faster. I think the if-else blocks may work fine because they are static, i.e. won't rely on traced values. It might be worth adding a `jit` arg to the test to see if this works. I'm also okay with addressing this in a separate PR instead.

Reply:
let's worry about jit elsewhere